crates/executor/src/blockchain_tree/block_indices.rs

```diff
@@ -0,0 +1,284 @@
+    pub fn new(
+        last_finalized_block: BlockNumber,
+        num_of_additional_canonical_block_hashes: u64,
+        canonical_chain: BTreeMap<BlockNumber, BlockHash>,
+    ) -> Self {
+        Self {
+            last_finalized_block,
+            num_of_additional_canonical_block_hashes,
+            fork_to_child: Default::default(),
+            canonical_chain,
+            blocks_to_chain: Default::default(),
+            index_number_to_block: Default::default(),
+        }
+    }
+
+    /// Return number of additional canonical block hashes that we need
+    /// to have to be able to have enought information for EVM execution.
+    pub fn num_of_additional_canonical_block_hashes(&self) -> u64 {
+        self.num_of_additional_canonical_block_hashes
+    }
+
+    /// Return fork to child indices
+    pub fn fork_to_child(&self) -> &HashMap<BlockHash, HashSet<BlockHash>> {
+        &self.fork_to_child
+    }
+
+    /// Return block to chain id
+    pub fn blocks_to_chain(&self) -> &HashMap<BlockHash, BlockChainId> {
+        &self.blocks_to_chain
+    }
+
+    /// Returns `true` if the Tree knowns the block hash.
+    pub fn contains_block_hash(&self, block_hash: BlockHash) -> bool {
+        self.blocks_to_chain.contains_key(&block_hash)
+    }
+
+    /// Check if block hash belongs to canonical chain.
+    pub fn is_block_hash_canonical(&self, block_hash: &BlockHash) -> bool {
+        self.canonical_chain.range(self.last_finalized_block..).any(|(_, &h)| h == *block_hash)
+    }
+
+    /// Last finalized block
+    pub fn last_finalized_block(&self) -> BlockNumber {
+        self.last_finalized_block
+    }
+
+    /// Insert block to chain and fork child indices of the new chain
+    pub fn insert_chain(&mut self, chain_id: BlockChainId, chain: &Chain) {
+        for (number, block) in chain.blocks().iter() {
+            // add block -> chain_id index
+            self.blocks_to_chain.insert(block.hash(), chain_id);
+            // add number -> block
+            self.index_number_to_block.entry(*number).or_default().insert(block.hash());
+        }
+        let first = chain.first();
+        // add parent block -> block index
+        self.fork_to_child.entry(first.parent_hash).or_default().insert(first.hash());
+    }
+
+    /// Get the chain ID the block belongs to
+    pub fn get_blocks_chain_id(&self, block: &BlockHash) -> Option<BlockChainId> {
+        self.blocks_to_chain.get(block).cloned()
+    }
+
+    /// Update all block hashes. iterate over present and new list of canonical hashes and compare
+    /// them. Remove all missmatches, disconnect them and return all chains that needs to be
+    /// removed.
+    pub fn update_block_hashes(
+        &mut self,
+        hashes: BTreeMap<u64, BlockHash>,
+    ) -> BTreeSet<BlockChainId> {
+        let mut new_hashes = hashes.iter();
+        let mut old_hashes = self.canonical_chain().iter();
+
+        let mut remove = BTreeMap::new();
```

Comment:
these are unique, right?
so can be a `Vec` instead?

Reply:
Right, and `Vec` makes code a little nicer.